File: src/main/java/com/wanted/preonboarding/ticket/presentation/PerformanceController.java
package com.wanted.preonboarding.ticket.presentation;

import com.wanted.preonboarding.core.domain.response.ApiResponse;
import com.wanted.preonboarding.ticket.application.PerformanceService;
import com.wanted.preonboarding.ticket.application.dto.response.PerformanceResponse;
import com.wanted.preonboarding.ticket.presentation.dto.request.AwaitRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
@RequestMapping("/api/v1/performances")
@RestController
public class PerformanceController {

    public static final String AWAIT_MESSAGE_FORMAT = "[%s] 예약 알림이 등록되었습니다.";
    private static final String FIND_PERFORMANCE_DEFAULT = "enable";

    private final PerformanceService performanceService;

    @GetMapping
    public ResponseEntity<ApiResponse<List<PerformanceResponse>>> findPerformances(
            @RequestParam(defaultValue = FIND_PERFORMANCE_DEFAULT) String isReserve
    ) {
        return ResponseEntity.ok(
                ApiResponse.ok(
                        performanceService.findPerformances(isReserve)
                )
        );
    }

    @GetMapping("/{performanceId}")
    public ResponseEntity<ApiResponse<PerformanceResponse>> findOne(@PathVariable UUID performanceId) {
        return ResponseEntity.ok(
                ApiResponse.ok(
                        performanceService.findOne(performanceId)
                )
        );
    }

    @PostMapping("/{performanceId}/await")
    public ResponseEntity<ApiResponse<String>> await(@PathVariable UUID performanceId,
                                                     @RequestBody @Validated AwaitRequest request
    ) {
        performanceService.await(performanceId, request.userId());
        return ResponseEntity.ok(ApiResponse.ok(String.format(AWAIT_MESSAGE_FORMAT, performanceId)));
    }

}

File: src/main/kotlin/com/wanted/preonboarding/ticket/presentation/PerformanceController.kt
package com.wanted.preonboarding.ticket.presentation

import com.wanted.preonboarding.core.CursorResult
import com.wanted.preonboarding.ticket.application.ticket.PerformanceService
import com.wanted.preonboarding.ticket.domain.PerformanceId
import com.wanted.preonboarding.ticket.presentation.common.ApiResponse
import com.wanted.preonboarding.ticket.presentation.response.PerformanceResponse
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.UUID

@RestController
class PerformanceController(
    private val performanceService: PerformanceService,
) {
    @GetMapping("/v1/performances/{performanceId}")
    fun getPerformance(
        @PathVariable performanceId: UUID,
    ): ApiResponse<PerformanceResponse> {
        val performance = performanceService.findPerformance(PerformanceId(performanceId))
        return ApiResponse.success(PerformanceResponse.from(performance))
    }

    @GetMapping("/v1/performances")
    fun findPerformances(
        @RequestParam reserveAvailable: Boolean,
        @RequestParam size: Int,
        @RequestParam(required = false) cursor: UUID?,
    ): ApiResponse<CursorResult<PerformanceResponse>> {
        val performances =
            performanceService.findAllPerformance(
                reserveAvailable,
                size,
                cursor?.let { PerformanceId(it) },
            )

        return ApiResponse.success(
            CursorResult(
                cursor = performances.cursor,
                hasNext = performances.hasNext,
                item = performances.item.map { PerformanceResponse.from(it) },
            ),
        )
    }
}
